src/library/scala/concurrent/Future.scala

```diff
@@ -566,18 +567,29 @@ object Future {
    */
   object never extends Future[Nothing] {
 
-    private[this] final val notGoingToHappen = new CountDownLatch(1)
-
     @throws[TimeoutException]
     @throws[InterruptedException]
     override final def ready(atMost: Duration)(implicit permit: CanAwait): this.type = {
       import Duration.{Undefined, Inf, MinusInf}
       atMost match {
         case u if u eq Undefined => throw new IllegalArgumentException("cannot wait for Undefined period")
-        case `Inf`               => notGoingToHappen.await()
+        case `Inf`               =>
+          while(!Thread.interrupted()) {
+            LockSupport.park(this)
+          }
+          throw new InterruptedException
         case `MinusInf`          => // Drop out
-        case f: FiniteDuration   =>
-          if (f > Duration.Zero) notGoingToHappen.await(f.toNanos, TimeUnit.NANOSECONDS)
+        case f: FiniteDuration if f > Duration.Zero  =>
+          var now = System.nanoTime()
+          val deadline = now + f.toNanos
+          while((deadline - now) > 0) {
+            LockSupport.parkNanos(this, deadline - now)
```

Comment:
Overflow safe: `deadline - now > 0`  ✅

Reply:
@retronym Yes—one needs to be careful with overflow—especially when it has to do with these things ^^